### ISSUE TITLE
Cannot get optimizer.Lamb to work 

### ISSUE THREAD

**Issue by sck-at-ucy** [stockeh/mlx-optimizers#36]
I have tried several of the provided optimizers without a problem, but I do not seem able to get Lamb to work: 

```
    optimizer = optim.Lamb(
        learning_rate=1e-4,  # Adjusted learning rate
        betas=(0.9, 0.999),
        weight_decay=0.01,  # Adjusted weight decay
        clamp_value=1000,  # Adjusted clamp value
        eps=1e-4  # Increased epsilon for numerical stability
    )
```

I keep getting nan for training loss despite experimenting with various values for the keyword params.

**Comment by sck-at-ucy** [comment]
This modified version solved the nan problem for me, so I am sharing it here: 

```
from typing import Callable, List, Union
import mlx.core as mx
from mlx.optimizers import Optimizer

class Lamb(Optimizer):
    r"""Layerwise Adaptive Large Batch Optimization [1].

    .. math::

        m_0 &= 0, v_0 = 0 \\
        m_t &= \beta_1 m_{t-1} + (1 - \beta_1) g_t \\
        v_t &= \beta_2 v_{t-1} + (1 - \beta_2) g_t^2 \\
        m_t &= m_t / (1 - \beta_1^t) \\
        v_t &= v_t / (1 - \beta_2^t) \\
        r_t &= \frac{m_t}{\sqrt{v_t} + \epsilon} \\
        \theta_{t+1} &= \theta_t - \eta \frac{\phi(\|\theta_t\|)}{\|r_t + \lambda \theta_t\|} \left(r_t + \lambda \theta_t\right)

    [1] You, Yang, et al., 2019. Large Batch Optimization for Deep Learning:
    Training BERT in 76 Minutes.
    https://arxiv.org/abs/1904.00962
    https://github.com/tensorflow/addons/blob/master/tensorflow_addons/optimizers/lamb.py

    Args:
        learning_rate (float or callable): learning rate :math:`\eta`.
        betas (Tuple[float, float], optional): coefficients
            :math:`(\beta_1, \beta_2)` used for computing running averages of the
            gradient and its square. Default: ``(0.9, 0.999)``
        eps (float, optional): term :math:`\epsilon` added to the
            denominator to improve numerical stability. Default: ``1e-6``
        weight_decay: weight decay. Default: ``0.9``
        clamp_value: clamp weight_norm in (clamp_value, inf) to scale trust ratio.
            Best to set to a high value (e.g :math:`10e3`). Default: ``10``
        adam: always use trust ratio = 1, which turns this
            into Adam. Useful for comparison purposes. Default: ``False``
        debias: debias adam by (1 - beta**step). Default: ``False``

    ..
    """

    def __init__(
        self,
        learning_rate: Union[float, Callable[[mx.array], mx.array]],
        betas: List[float] = [0.9, 0.999],
        weight_decay: float = 0.0,
        clamp_value: float = 10,
        eps: float = 1e-6,
        adam: bool = False,
        debias: bool = False,
    ):
        super().__init__()

        self._maybe_schedule("learning_rate", learning_rate)
        self.betas = betas
        self.weight_decay = weight_decay
        self.clamp_value = clamp_value
        self.eps = eps
        self.adam = adam
        self.debias = debias

    def init_single(self, parameter: mx.array, state: dict):
        """Initialize optimizer state"""
        state["m"] = mx.zeros_like(parameter)
        state["v"] = mx.zeros_like(parameter)

    def apply_single(self, gradient: mx.array, parameter: mx.array, state: dict):
        """Performs a single optimization step, updating :math:`m` and :math:`v`"""
        lr = self.learning_rate.astype(gradient.dtype)
        b1, b2 = self.betas

        m = state["m"]
        v = state["v"]

        # Update first and second moments
        m = b1 * m + (1 - b1) * gradient
        v = b2 * v + (1 - b2) * mx.square(gradient)

        # Bias correction if debias is True
        bias_correction = (
            max(mx.sqrt(1 - mx.power(b2, self.state["step"])) / (1 - mx.power(b1, self.state["step"])), self.eps)
            if self.debias else 1
        )

        eta = lr * bias_correction

        # Compute weight norm with clipping
        weight_norm = mx.clip(mx.linalg.norm(parameter), self.clamp_value, 1e5)  # Replace inf with large value

        # Calculate adam step with added epsilon for stability
        adam_step = m / (mx.sqrt(v) + self.eps)

        if self.weight_decay:
            adam_step = adam_step + self.weight_decay * parameter

        # Compute the norm of adam_step with epsilon for numerical stability
        adam_norm = mx.linalg.norm(adam_step) + self.eps  # Avoid zero in denominator
        trust_ratio = 1 if self.adam else weight_norm / adam_norm

        # Update state for momentum terms
        state["m"] = m
        state["v"] = v

        # Parameter update
        return parameter - eta * trust_ratio * adam_step
```

**Comment by stockeh** [comment]
@sck-at-ucy I appreciate the tests! I made a pull request (#39) that better aligns Lamb with the underlying implementation. I believe this will be better than clipping the values. I'll get a build out soon! Feel free to open another issue if you notice other issues 🙂

**Comment by sck-at-ucy** [comment]
Thank you, I will test the new implementation
